File: NQueens.py
import time

class NQueens:
    def __init__(self, n):
        self.n = n
        self.board = [-1] * n
        self.cols = set()
        self.diagonals = set()
        self.antiDiagonals = set()
        self.solutions = []

    def is_safe(self, row, col):
        return col not in self.cols and \
               (row - col) not in self.diagonals and \
               (row + col) not in self.antiDiagonals

    def solve(self, row):
        if row == self.n:
            self.solutions.append(self.board.copy())
            return True
        for col in range(self.n):
            if self.is_safe(row, col):
                self.board[row] = col
                self.cols.add(col)
                self.diagonals.add(row - col)
                self.antiDiagonals.add(row + col)
                if self.solve(row + 1):
                    return True
                self.cols.remove(col)
                self.diagonals.remove(row - col)
                self.antiDiagonals.remove(row + col)
        return False

    def solve_n_queens(self):
        self.solve(0)
        return self.solutions

def display_board(board):
    for row in board:
        line = ['.' for _ in range(len(board))]
        line[row] = 'Q'
        print(''.join(line))
    print('-' * len(board) * 2)

if __name__ == "__main__":
    n = 25
    start_time = time.time()  # Start time
    solver = NQueens(n)
    solutions = solver.solve_n_queens()
    end_time = time.time()  # End time

    if solutions:
        print(f"Found a solution for {n}-Queens problem in python:")
        display_board(solutions[0])
    else:
        print(f"No solutions found for {n}-Queens problem.")

    print(f"Execution time: {end_time - start_time} seconds")

File: NQueens.cpp
#include <iostream>
#include <vector>
#include <set>
#include <chrono>

class NQueens {
    int n;
    std::vector<int> board;
    std::set<int> cols, diagonals, antiDiagonals;
    std::vector<std::vector<int>> solutions;

public:
    NQueens(int n) : n(n), board(n, -1) {}

    bool isSafe(int row, int col) {
        return cols.find(col) == cols.end() &&
               diagonals.find(row - col) == diagonals.end() &&
               antiDiagonals.find(row + col) == antiDiagonals.end();
    }

    bool solve(int row) {
        if (row == n) {
            solutions.push_back(board);
            return true;
        }
        for (int col = 0; col < n; col++) {
            if (isSafe(row, col)) {
                board[row] = col;
                cols.insert(col);
                diagonals.insert(row - col);
                antiDiagonals.insert(row + col);
                if (solve(row + 1)) {
                    return true;
                }
                cols.erase(col);
                diagonals.erase(row - col);
                antiDiagonals.erase(row + col);
            }
        }
        return false;
    }

    std::vector<std::vector<int>> solveNQueens() {
        solve(0);
        return solutions;
    }

    static void displayBoard(const std::vector<int>& board) {
        for (int row : board) {
            for (int i = 0; i < board.size(); i++) {
                std::cout << (i == row ? 'Q' : '.');
            }
            std::cout << std::endl;
        }
        std::cout << std::string(board.size() * 2, '-') << std::endl;
    }
};

int main() {
    int n = 25;
    auto startTime = std::chrono::high_resolution_clock::now();
    NQueens solver(n);
    auto solutions = solver.solveNQueens();
    auto endTime = std::chrono::high_resolution_clock::now();

    if (!solutions.empty()) {
        std::cout << "Found a solution for " << n << "-Queens problem C++ edition:" << std::endl;
        NQueens::displayBoard(solutions[0]);
    } else {
        std::cout << "No solutions found for " << n << "-Queens problem." << std::endl;
    }
    std::chrono::duration<double> elapsedTime = endTime - startTime;
    std::cout << "Execution time: " << elapsedTime.count() << " seconds" << std::endl;

    return 0;
}
